mm2src/coins/utxo/utxo_common.rs

```diff
@@ -1624,6 +1627,43 @@ pub fn my_address<T: UtxoCommonOps>(coin: &T) -> Result<String, String> {
     }
 }
 
+/// Hash message for signature using Bitcoin's message signing format.
+/// sha256(sha256(PREFIX_LENGTH + PREFIX + MESSAGE_LENGTH + MESSAGE))
+pub fn sign_message_hash(coin: &UtxoCoinFields, message: &str) -> Option<[u8; 32]> {
+    let message_prefix = coin.conf.sign_message_prefix.clone()?;
+    let mut stream = Stream::new();
+    let prefix_len = CompactInteger::from(message_prefix.len());
+    prefix_len.serialize(&mut stream);
+    stream.append_slice(message_prefix.as_bytes());
+    let msg_len = CompactInteger::from(message.len());
+    msg_len.serialize(&mut stream);
+    stream.append_slice(message.as_bytes());
+    Some(dhash256(&stream.out()).take())
+}
+
+pub fn sign_message(coin: &UtxoCoinFields, message: &str) -> SignatureResult<String> {
+    let message_hash = sign_message_hash(coin, message).ok_or(SignatureError::PrefixNotFound)?;
+    let private_key = coin.priv_key_policy.key_pair_or_err()?.private();
+    let signature = private_key.sign_compact(&H256::from(message_hash))?;
+    Ok(base64::encode(&*signature))
+}
+
+pub fn verify_message<T: UtxoCommonOps>(
+    coin: &T,
+    signature_base64: &str,
+    message: &str,
+    address: &str,
+) -> VerificationResult<bool> {
+    let message_hash = sign_message_hash(coin.as_ref(), message).ok_or(VerificationError::PrefixNotFound)?;
+    let signature = CompactSignature::from(base64::decode(signature_base64)?);
+    let pubkey = Public::recover_compact(&H256::from(message_hash), &signature)?;
+    let address_from_pubkey = coin
+        .address_from_pubkey(&pubkey)
```

Comment:
I doubt that it will work properly if coin uses Segwit address format, but `verify_message` is called for legacy address. Could you add a test case for this and fix if it fails, please?

Reply:
Added a test case, everything works fine for the Segwit address format. There was a slight problem with the `utxo_coin_fields_for_test ` function that it wasn't really creating a Segwit address, also fixed it.

Reply:
Replied in https://github.com/KomodoPlatform/atomicDEX-API/pull/1256#discussion_r863559498